draft-ietf-quic-transport.md

```diff
@@ -2070,91 +2069,107 @@ following layout:
+This stateless reset design is specific to QUIC version 1.  An endpoint that
 supports multiple versions of QUIC needs to generate a stateless reset that will
-be accepted by clients that support any version that the server might support
+be accepted by peers that support any version that the endpoint might support
 (or might have supported prior to losing state).  Designers of new versions of
 QUIC need to be aware of this and either reuse this design, or use a portion of
 the packet other than the last 16 octets for carrying data.
 
 
 #### Detecting a Stateless Reset
 
-A client detects a potential stateless reset when a packet with a short header
-either cannot be decrypted or is marked as a duplicate packet.  The client then
-compares the last 16 octets of the packet with the Stateless Reset Token
-provided by the server in its transport parameters.  If these values are
-identical, the client MUST enter the draining period and not send any further
-packets on this connection.  If the comparison fails, the packet can be
-discarded.
+An endpoint detects a potential stateless reset when a packet with a short
+header either cannot be decrypted or is marked as a duplicate packet.  The
+endpoint then compares the last 16 octets of the packet with the Stateless Reset
+Token provided by its peer, either from the NEW_CONNECTION_ID frame or the
+server transport parameters.  If these values are identical, the endpoint MUST
+enter the draining period and not send any further packets on this connection.
+If the comparison fails, the packet can be discarded.
 
 
 #### Calculating a Stateless Reset Token
 
 The stateless reset token MUST be difficult to guess.  In order to create a
-Stateless Reset Token, a server could randomly generate {{!RFC4086}} a secret
+Stateless Reset Token, an endpoint could randomly generate {{!RFC4086}} a secret
 for every connection that it creates.  However, this presents a coordination
-problem when there are multiple servers in a cluster or a storage problem for a
-server that might lose state.  Stateless reset specifically exists to handle the
-case where state is lost, so this approach is suboptimal.
+problem when there are multiple instances in a cluster or a storage problem for
+a endpoint that might lose state.  Stateless reset specifically exists to handle
+the case where state is lost, so this approach is suboptimal.
 
 A single static key can be used across all connections to the same endpoint by
 generating the proof using a second iteration of a preimage-resistant function
-that takes three inputs: the static key, the server's connection ID (see
-{{connection-id}}), and an identifier for the server instance.  A server could
-use HMAC {{?RFC2104}} (for example, HMAC(static_key, server_id ||
+that takes three inputs: the static key, the connection ID chosen by the
+endpoint (see {{connection-id}}), and an instance identifier.  An endpoint could
+use HMAC {{?RFC2104}} (for example, HMAC(static_key, instance_id ||
```

Comment:
IIUC, incorporating an instance_id doesn't work for connections that are diverted to a new instance, which is a common case, possibly even the most common case. Server restarts are unlikely to be a common use case, since usually servers are drained before restarting, and server crashes are not frequent. Short-term routing flaps however are common.  I think the text should say that for Stateless Reset to work across instances, the HMAC could exclude the instance_id. (Correct me if my understanding is wrong.)

Reply:
I've addressed that below.  But I've expanded on it a little in my latest changes.  The trick here is that you can recover the instance ID using the connection ID.  I will follow up with a change that includes the fixes we discussed.  Part of that change involves removing the instance ID from this.

Reply:
And of course, the issue is that if you enable servers to SR after a routing flap, you also enable those servers to be used as oracles.  The only defense is to make it impossible for the routing to be affected by the attacker.

Reply:
@martinthomson : Ah, right, I see it's addressed below. Given this is an example construction, I would suggest making that explicitly clear up at the beginning of this paragraph.
@MikeBishop : Yes, that is true, but defending the routing infrastructure against attacks seems out of scope for us, since similar attacks can be launched against TCP as well, where you'd receive an RST or ICMP unreachable.

Reply:
Defending it is out of scope.  Noting that you create an attack vector if you fail to defend it yourself, however, bears mentioning.